### ISSUE TITLE
Build Issues: Assets file 'C:\DynamicsValue\fake-xrm-easy\src\FakeXrmEasy\obj\project.assets.json' doesn't have a target for 'net452'.

### ISSUE THREAD

**Issue by bwmodular** [DynamicsValue/fake-xrm-easy#55]
I'm trying to build the project(s) using the instructions in Readme.md. 
I have cloned all of the 2x-dev repos to a folder structure like this:
![image](https://user-images.githubusercontent.com/1889586/188403524-a120f27d-be7b-48a8-b169-e74f908cc6f8.png)
I have all of the prerequisites installed.
When I run any of the build scripts (eg _./build.ps1 -targetFrameworks all_) I get a slew of errors, but the first is this:
_C:\Program Files\dotnet\sdk\6.0.101\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file '_C:\DynamicsValue\fake-xrm-easy\src\FakeXrmEasy\obj\project.assets.json' doesn't have a target for 'net452'_. Ensure that restore has run and that you have included 'net452' in the TargetFrameworks for your project. [C:\DynamicsValue\fake-xrm-easy\src\FakeXrmEasy\FakeXrmEasy.csproj]_

Not sure if I'm doing something wrong?

**Comment by BetimBeja** [comment]
Hi @bwmodular 
I think you need to `build-push-local.ps1 -targetFrameworks all` in the following order https://github.com/DynamicsValue/fake-xrm-easy#build-dependencies

**Comment by jordimontana82** [comment]
It could be what Betim says, you need the projects in the order specified. If you try building just the fake-xrm-easy-abstractions repo first to try? @bwmodular ?

**Comment by jordimontana82** [comment]
I'm assuming you also have the sdks for net 4.5.2 and 4.6.2 installed?

**Comment by jordimontana82** [comment]
@bwmodular  there is also the build-all.ps1 script which tries to build everything too, it might take quite some time, but it's a once off task, then we can push versions from each repo.

![imagen](https://user-images.githubusercontent.com/3968950/188457860-54df4db1-7a3a-465f-957c-579f013cb677.png)


Please run another git-checkout.ps1 script to sync all branches in all repos to 2x-dev before running the build-all one. 

 That's something I added after the README.md  ... too many commits!

**Comment by bwmodular** [comment]
Thanks both - running things in the right order is definitely helping - Sorry! Still getting some errors but I'll spend some more time seeing how much further I can get and report back....
I did try build-all.ps1 first but that really wasn't happy at all, so trying each project one by one now.....

**Comment by bwmodular** [comment]
Well, I tried, but still no luck getting anything built or pushed.
This all makes me feel pretty dumb as I'm sure it must be something obvious.

Anyway, it's worth persevering because:
a.) I want to be able to contribute to the project and 
b.) this needs to be made as easy and foolproof as possible - one of the great things about v1 was that you could just download and compile it and then you could see and hundreds of examples of how to write tests to test certain things, whereas now it seems like it might take tens of minutes just to get the point where everything is built. You don't want people giving up just because it's too complicated to get the project to build.

I tried doing build-all target-frameworks all - this is a mixed bag of failure and successes while the script runs but ultimately falls over with this error:
FakeXrmEasy.Plugins.Performance\PipelineBenchmarks.cs(39,21): error CS0117: 'PluginStepDefinition' does not contain a definition for 'ImagesDefinitions'

So I then tried to break it down solution by solution.

Here are my results after running _build-push-local.ps1 -targetFrameworks all_ in the order specified in the dependencies section.
My forks were only created and cloned yesterday but I’ve made sure they are up to date with orig.
**Abstractions:**
Running gives me this smiley result:
 
![image](https://user-images.githubusercontent.com/1889586/188516313-d5bf2abe-8ac5-45c1-9012-5c2caa3a9804.png)

but while the script is running there are lots of errors thrown up:
 
![image](https://user-images.githubusercontent.com/1889586/188516325-ddfdb883-1ef2-47c6-ba85-a60dfe40f152.png)

and
 
![image](https://user-images.githubusercontent.com/1889586/188516342-84915b7d-520f-4eaa-b08c-e12871fb9750.png)

Trying to build the Abstractions solution in Visual Studio straight afterwards gives me loads of errors like this:
 
![image](https://user-images.githubusercontent.com/1889586/188516359-86f108a7-5f3e-4e58-a822-db4a7616b85f.png)

**Core**
Again, seems like a happy ending:
![image](https://user-images.githubusercontent.com/1889586/188516363-0380f775-37dc-4577-a80e-017d7261f177.png)

 
But lots of errors along the way:
 
![image](https://user-images.githubusercontent.com/1889586/188516372-510ffed9-9e7f-4242-815d-05f6c73662ff.png)

And if I try and build in Visual Studio:
![image](https://user-images.githubusercontent.com/1889586/188516378-875fbfeb-6ac7-48d0-96e3-132de4013911.png)

 
**Messages** – Pretty much the same as Core
**Plugins** – Lots of errors while the script runs and then a hard failure:
C:\DynamicsValue\fake-xrm-easy-plugins\performance\FakeXrmEasy.Plugins.Performance\PipelineBenchmarks.cs(39,21): error CS0117: 'PluginStepDefinition' does not contain a definition for 'ImagesDefinitions' 
**Code-Activities** – The script runs without errors but I still can't build in Visual Studio afterwards
**Fake-xrm-easy** – fails with the original error about not having a target for net452, along with lots of not surprising errors about missing dependencies.
![image](https://user-images.githubusercontent.com/1889586/188516396-ae09f519-bc2d-4072-b05f-a09145b351b8.png)


Bottom line is that I couldn't really get anything building and certainly not getting to the point where anything compiles in Visual Studio.

**Comment by jordimontana82** [comment]
Hi Ben, Agree with all you said.

Maybe it could be that you pulled a version while I was in the middle of some changes in the package dependencies in the .csproj, cause I see a mixture of 2.2.0 and 2.1.1 in your logs when building the abstractions package.

If you do 

    pwsh ./git-pull.ps1

 in the abstractions package to see abstractions package errors go away anyway?

**Comment by jordimontana82** [comment]
@bwmodular  I've raised another issue to improve the whole cloning, pulling , and building for the first time:

#56 

@BetimBeja 

Any ideas welcome :)

**Comment by bwmodular** [comment]
> 
> If you do
> 
> ```
> pwsh ./git-pull.ps1
> ```
> 
> in the abstractions package and then build again to see abstractions package errors go away now?

@jordimontana82 Thanks for this. I think I'm all up to date:
![image](https://user-images.githubusercontent.com/1889586/188580725-5abc9bc5-5458-47d8-8101-9bdc8a2374f2.png)
and
![image](https://user-images.githubusercontent.com/1889586/188581356-9fbbbf3c-02b0-4287-b3a7-12e89e1136e3.png)

**Comment by jordimontana82** [comment]
@bwmodular  You can run git-pull.ps1 without params, and it'll pull all branches.

**Comment by jordimontana82** [comment]
Does it build now?

**Comment by bwmodular** [comment]
gitppull on fake-xrm-easy & fake-xrm-easy-abstractions shows I'm up to date:
![image](https://user-images.githubusercontent.com/1889586/188596125-fe19af79-110a-45bb-8eab-11319e4c4cc8.png)
build-push-local.ps1 -targetFrameworks all still producing lots of errors.
I suspect that we're going to have to start going through the errors from the top and try and work out each one, but I don't want to take your time this week as you're preparing for Bizz. Maybe we can have a teams session after Bizz so you can see what's going on.
This is the first error that gets thrown up:
_C:\Program Files\dotnet\sdk\6.0.101\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file 'C:\DynamicsValue\fake-xrm-easy-abstractions\src\FakeXrmEasy.Abstractions\obj\project.assets.json' doesn't have a target for 'net462'. Ensure that restore has run and that you have included 'net462' in the TargetFrameworks for your project. [C:\DynamicsValue\fake-xrm-easy-abstractions\src\FakeXrmEasy.Abstractions\FakeXrmEasy.Abstractions.csproj]_
This comes after a succesful build and pack of FakeXrmEasy.Abstractions.v2016,so I think its FAKE_XRM_EASY_365 that's failing. All subsequent builds of the different Abstractions packages fail after this one.

**Comment by jordimontana82** [comment]
+1 to Teams session

**Comment by bwmodular** [comment]
Thought I'd try a different approach.
Abstractions should build on its own as it's at the top of the dependency hirarchy, yes?

I downloaded the current version of the Abstractions solution from orig:
https://github.com/DynamicsValue/fake-xrm-easy-abstractions 
just using the Download Zip option from GitHub web. So I must have the latest version.

Opened in Visual Studio, tried to build.
Get lots of these:
![image](https://user-images.githubusercontent.com/1889586/188655479-cbbe1517-2cbb-4bff-8189-4c1913fe1857.png)

Go to 'Manage NuGet Packagesfor Solution'  and these are the only packages it thinks are installed:
![image](https://user-images.githubusercontent.com/1889586/188655720-f7b28c2d-969b-46d8-a3b6-0facd9a34531.png)

So not surprising it can't find Xrm or any of the other references.
Tried it on two different machines with 4.5.2 and 4.62 intalled.

It must be something obvious, but can't work out what I'm missing.

**Comment by jordimontana82** [comment]
> Thought I'd try a different approach.
Abstractions should build on its own as it's at the top of the dependency hierarchy, yes?

Yep

That's really weird, the restore should just work with build.ps1 -targetFrameworks all (there is a step that restores all packages there). 

Not sure if related, but I didn't try to build the project with .net 6.0 yet, only with net core 3.1. I don't know if you could try, otherwise I'll try building it in .net60 when I have a chance.

**Comment by bwmodular** [comment]
Interestingly, the 462 build works OK, but the .netCore one doesn't.
So this works fine: .\build.ps1 -targetFrameworks net462
But this .\build.ps1 -targetFrameworks netcoreapp3.1
does not, giving lots of:  _The type or namespace name 'EntityReference' could not be found_ type errors.

This would (kind of) explain why I can't build in Visual Studio. In Visual Studio, I can't actually change the TargetFramework, it's greyed out:
![image](https://user-images.githubusercontent.com/1889586/188683504-3cc20c85-2f2d-4a6f-9770-9bd2da9edb27.png)

.Net Core SDK 3.1 is definitely installed:
![image](https://user-images.githubusercontent.com/1889586/188683833-72de866a-f8be-45bb-90dd-8be91c66b704.png)

**Comment by jordimontana82** [comment]
HI Ben,

netcoreapp3.1 won't build in 2x-* branches, you need to build it in 3x-* ones.  However ./build.ps1 -targetFrameworks all should work in all.

**Comment by bwmodular** [comment]
Yes of course - obvious really, sorry.

So, _.\build.ps1 -targetFrameworks net462_ works
and _.\build.ps1 -targetFrameworks all_ works
but _.\build-push-local.ps1 -targetFrameworks all_ fails with the missing references errors
and Visual Studio won't build at all.

**Comment by jordimontana82** [comment]
Ok, it seems this definitely needs a Teams call cause I ran all these scripts in both branches without issues myself at least :)

**Comment by bwmodular** [comment]
No luck deleting the packages folder - same issues. I'd have been surprised if it works as I've tried it on two different machines and both via cloned repo and direct download from GitHub into a different folder. In Visual Studio, Restore Nuget Packages says 'All Packages up to date' and the packages folder is full of folders including crm sdk assemblies, but in Visual Studio it looks like none of the NuGet packages are installed. All v odd....

**Comment by jordimontana82** [comment]
What version of VS and Powershell are you using Ben?

**Comment by bwmodular** [comment]
I've tried it in VS 2019 and VS 2022.
But Powershell 5.1 - Hmm - I'll try upgrading to v7. I assumed that my new win11 laptop would have latest PS but no. I'll upgrade and try again....

**Comment by jordimontana82** [comment]
I'm using a Win10 and Ubuntu myselft, so didnt' try Win 11 to at least build this project...but good to know Win 11 comes with a fancy PS 5 by default!

**Comment by bwmodular** [comment]
Upgrading to PS 7 on my Win10 machine hasn't helped.
I'm still getting lots of errors - mostly the old favourite: Assets file 'C:\DynamicsValue\fake-xrm-easy-abstractions\src\FakeXrmEasy.Abstractions\obj\project.assets.json' doesn't have a target for 'net462'.
And (obviously) still no luck with Visual Studio.
I doubt my Win11 machine will be any different but will report back if I'm proved wrong :-)

**Comment by jordimontana82** [comment]
Are you running scripts as a local admin? Wondering about file permissions and so on....

**Comment by bwmodular** [comment]
Yes, you can't run most of these commands in PS without having first of all given PS permission to download files, etc, and I've also tried in ISE as Admin.

**Comment by bwmodular** [comment]
Well, I worked out why the abstractions project wouldn't build in Visual Studio at least.
When you first open a project in VS it defaults the Solution Configuration to debug.
![image](https://user-images.githubusercontent.com/1889586/188937005-49ef0cf3-2284-45d3-a6ba-71292dcc5f6a.png)
Changing it to one of the FakeXrmEasy configs causes it to download the appropriate libraries from NuGet.
But you still see this error when you build:
![image](https://user-images.githubusercontent.com/1889586/188935648-125f5842-3a90-4b01-b07c-8d74da21402e.png)
But once you close and re-open the solution it builds ok!

Well, not every configuration does, but most do.
If I switch to FAKE_XRM_EASY_365, I get this error:
![image](https://user-images.githubusercontent.com/1889586/188936329-2a91e4ec-fc03-4a87-ab7b-47bd97c6bd6a.png)
and this:
![image](https://user-images.githubusercontent.com/1889586/188936407-de32ae8e-4561-456d-86ae-2041e674b767.png)
and if I open FAKE_XRM_EASY_9 I get this:
![image](https://user-images.githubusercontent.com/1889586/188936825-d1b96b72-d00c-4d71-b20a-1ad095e491a8.png)

So, I can now get every version apart from 9 and 365 to build in VS.
Still no luck with build.ps though, but I feel like I've made a small amount of progress today :-)

**Comment by jordimontana82** [comment]
Ahhh ok. Yeah, if try closing & restarting VS after changing the configuration?

**Comment by jordimontana82** [comment]
The project builds just fine via console, but VS seems to go mad with these conditional symbols in the .csproj...

**Comment by bwmodular** [comment]
If I amend this line in the two csproj files (src & test) in Abstractions, I can get all of the projects to build in VS:
<TargetFrameworks Condition="'$(Configuration)'=='FAKE_XRM_EASY_365'">net462;net452</TargetFrameworks>
to
<TargetFrameworks Condition="'$(Configuration)'=='FAKE_XRM_EASY_365'">net462</TargetFrameworks>
- removing the reference to net452.

I don't know if this is something I can legitimately do, but it at least allows me to build.

One really annoying thing in VS is that you can't even see which configuration you are currently using!
![image](https://user-images.githubusercontent.com/1889586/189089687-4dea9ee8-a956-4cdc-a1dc-75eaf24029f3.png)
ery annoying when you've having to switch between them to test different versions.

**Comment by jordimontana82** [comment]
@bwmodular  There have a few updates to the build and clean scripts. Do you still have these issues by any chance (i.e. using Rider lately and the switching of configurations is seamless....)

**Comment by temmyraharjo** [comment]
Hi @jordimontana82,
Yes, I getting the same error using VS 2022. But when using Rider (sorry Jordi, I'm still using Windows 11), I'm not getting that error.
![Screenshot 2023-10-22 163930](https://github.com/DynamicsValue/fake-xrm-easy/assets/8713243/7277a1ae-63bd-4ab4-a624-cc56bbcf350c)

**Comment by jordimontana82** [comment]
@temmyraharjo based on your screenshot (which is really helpful!) it might be there are still Debug configs floating around in .sln files, like: 

https://github.com/DynamicsValue/fake-xrm-easy-messages/blob/2x-dev/FakeXrmEasy.Messages.sln#L21-L23

If you can try to remove those 3 lines from the .sln file to see if that fixes it for you? If so, I'll raise an issue to double check all repos.

**Comment by temmyraharjo** [comment]
Wow! Yes @jordimontana82, removing L21-L23 (Debug) and L42-L44 (Release) does fix the issue! 
![Screenshot 2023-10-23 100153](https://github.com/DynamicsValue/fake-xrm-easy/assets/8713243/e217f665-2713-4c00-bb76-e57db851e377)

**Comment by bwmodular** [comment]
Well that’s very exciting! I did try it again last week and did notice the problem was still there. Well done for finding the cause!

**Comment by jordimontana82** [comment]
Yeah! The error was misleading in that it seemed it was an issue with the project file, but it was the solution instead.
